design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+  {'name':'large','numThreads':2}
+]
+```
+
+  * `tserver.compaction.major.service.<service>.planner.opts.maxOpen`: number of files that will be included in a single compaction
+
+
+# External Compactions
+
+External compactions are major compactions (partial or full) that occur outside of the Tablet Server using the following set of components:
+
+## Components
+
+  * *CompactionCoordinator*: a component that maintains a global external compaction queue and schedules compactions for execution in a Compactor.
+
+  * *Compactor*: a process that asks the CompactionCoordinator for work and performs compactions one at a time for a specified external queue.
+  
+## Overview
+A CompactionService could be defined to use a Planner that has the ability to create internal and/or external CompactionExecutors. For example, let’s say that the configuration for this service is:
+
+```
+tserver.compaction.major.service.<service>.planner.opts.executors -
+[
+  {'name':'small','maxSize':'32M','numThreads':2},
+  {'name':'medium','maxSize':'128M','numThreads':2},
+  {'name':'large', ‘externalQueue’:’EQ1’}
+]
+```
+
+This planner would create CompactionExecutor’s for the internal queues and a new ExternalCompactionExecutor for the external queue. When ExternalCompactionExecutor::submit is called this Executor would eventually communicate summary information to the CompactionCoordinator that enables it to later request work for EQ1 from the tserver. The CompactionCoordinator adds this request to the list of requests from all TabletServers. Compactor processes that are running will request the next compaction task from the CompactionCoordinator when they are free to do work and will notify the CompactionCoordinator when the compaction has completed (or failed).
+
+## APIs
+
+### Compaction Coordinator RPC API
+
+```
+/**
+ * Called by tablet servers to get the status of compaction
+**/
+ExternalCompactionStatus[] getCompactionStatus(TabletId[])
+
+
+/**
+ * Called by tablet servers to get the number of running and queued compaction requests
+**/
+ExternalCompactionStats getCompactionStats(TabletId, String queueName)
+
+
+/**
+ * Called by compactor processes to get work to do for a queue.  The compactor Id can be used    to check if the process is alive. The implementation will find the top tablet in its queue and then    call reserveJob for that tablet on the tablet server.  This could be a blocking RPC call where it does not return until there is work, however not sure how well this will work.
+ */
+CompactionJob getCompactionJob(String queueName, CompactorId)
+
+
+/**
+ * Called by compactor process to indicate work is done. The implementation will call  finishedJob on the tablet servers.
+ **/
+void finishedJob(CompactionJob, CompactionStats)
+
+
+/**
+  * Called by compactor process to update the coordinator with the state of the running compaction
+ **/
+void updateCompactionStatus(CompactionJob, State, message, timestamp)
+```
+
```

Comment:
I feel like at least one of these methods should include a timeout, I am not sure where that would be determined though.